packages/tpetra/core/ext/TpetraExt_MatrixMatrix_def.hpp

```diff
@@ -3504,6 +3084,282 @@ merge_matrices(CrsMatrixStruct<Scalar, LocalOrdinal, GlobalOrdinal, Node>& Aview
+  typedef typename MMdetails::AddKernels<Scalar, LocalOrdinal, GlobalOrdinal, Node>::col_inds_array::non_const_type LView;
+  typedef typename Map<LocalOrdinal, GlobalOrdinal, Node>::local_map_type LocalMap;
+  GlobalOrdinal nentries = gids.extent(0);
+  typedef Kokkos::Bitset<device_type> bitset_t;
+  typedef Kokkos::ConstBitset<device_type> const_bitset_t;
+  //Determine the set of GIDs in the column map using a dense bitset (covers entire possible range)
+  bitset_t presentGIDs(maxCol - minCol + 1);
+  Kokkos::parallel_for(range_type(0, nentries), ColMapFunctors::GatherPresentEntries<GlobalOrdinal, GView, bitset_t>(minCol, gids, presentGIDs));
+  execution_space().fence();
+  const_bitset_t constPresentGIDs(presentGIDs);
+  GView mapGIDs("Sum colmap GIDs", constPresentGIDs.count());
+  std::cout << "Rank " << myRank << ": determined there are " << mapGIDs.extent(0) << " unique global cols on this proc.\n";
+  Kokkos::parallel_scan(range_type(0, maxCol - minCol + 1), ColMapFunctors::ListGIDs<GlobalOrdinal, GView, const_bitset_t>(minCol, mapGIDs, constPresentGIDs));
+  execution_space().fence();
+  //Construct the map
+  auto colMap = rcp(new Map<LocalOrdinal, GlobalOrdinal, Node>(
+        Teuchos::OrdinalTraits<Tpetra::global_size_t>::invalid(), //have Tpetra compute global #GIDs
+        mapGIDs, indexBase, comm));
+  //Get the device-friendly local map
+  LocalMap localColMap = colMap->getLocalMap();
+  //Use it to convert gids to lids
+  Kokkos::parallel_for(range_type(0, nentries), ColMapFunctors::ConvertGlobalToLocal<GlobalOrdinal, GView, LView, LocalMap>(gids, lids, localColMap));
+  execution_space().fence();
+  return colMap;
+}
+
+template<typename SC, typename LO, typename GO, typename NO>
+void AddKernels<SC, LO, GO, NO>::
+addSorted(
+  const typename MMdetails::AddKernels<SC, LO, GO, NO>::values_array& Avals,
+  const typename MMdetails::AddKernels<SC, LO, GO, NO>::row_ptrs_array_const& Arowptrs,
+  const typename MMdetails::AddKernels<SC, LO, GO, NO>::col_inds_array& Acolinds,
+  const typename MMdetails::AddKernels<SC, LO, GO, NO>::impl_scalar_type scalarA,
+  const typename MMdetails::AddKernels<SC, LO, GO, NO>::values_array& Bvals,
+  const typename MMdetails::AddKernels<SC, LO, GO, NO>::row_ptrs_array_const& Browptrs,
+  const typename MMdetails::AddKernels<SC, LO, GO, NO>::col_inds_array& Bcolinds,
+  const typename MMdetails::AddKernels<SC, LO, GO, NO>::impl_scalar_type scalarB,
+  typename MMdetails::AddKernels<SC, LO, GO, NO>::values_array& Cvals,
+  typename MMdetails::AddKernels<SC, LO, GO, NO>::row_ptrs_array& Crowptrs,
+  typename MMdetails::AddKernels<SC, LO, GO, NO>::col_inds_array& Ccolinds)
+{
+  using Teuchos::TimeMonitor;
+  TEUCHOS_TEST_FOR_EXCEPTION(Arowptrs.extent(0) != Browptrs.extent(0), std::runtime_error, "Can't add matrices with different numbers of rows.");
+  auto nrows = Arowptrs.extent(0) - 1;
```

Comment:
What if `Arowptrs.extent(0)` is zero?  We allow that case for a matrix with zero rows.